polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+      SmallVectorImpl<AccessorPtrPairType> &accessorPairs) const;
+  /// Version \p call.
+  void versionCall(CallOpInterface call) const;
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// FunctionSpecializationPass
+//===----------------------------------------------------------------------===//
+
+void FunctionSpecializationPass::runOnOperation() {
+  SmallVector<FunctionOpInterface> candidates;
+  getOperation()->walk([&](FunctionOpInterface func) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Processing function \"" << func.getName() << "\"\n");
+    if (isCandidateFunction(func))
+      candidates.push_back(func);
+  });
+
+  ModuleOp module = getOperation();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+
+  for (FunctionOpInterface func : candidates) {
+    FunctionOpInterface clonedFunc = cloneFunction(func);
+    setInnerDisjointAttribute(clonedFunc, isCandidateArg);
+
+    for (Operation *op : userMap.getUsers(func)) {
+      // Due to temporary condition to only allow function called directly by a
+      // GPU kernel.
+      assert(op->getParentOfType<gpu::GPUFuncOp>() &&
+             "Expecting calls only in GPU kernel");
+
+      auto call = cast<CallOpInterface>(op);
+      versionCall(call);
+      updateCallee(call, clonedFunc);
+    }
+    ++numFunctionSpecialized;
+  }
+}
+
+bool FunctionSpecializationPass::isCandidateFunction(
+    FunctionOpInterface func) const {
+  if (!isPotentialKernelBodyFunc(func)) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "not a candidate: not a potential kernel body function\n");
+    return false;
+  }
+
+  // Temporary condition to only allow function called directly by a GPU kernel.
+  // TODO: allow maximum depth of 2.
+  Optional<unsigned> maxDepth = getMaxDepthFromAnyGPUKernel(func);
+  assert(maxDepth.has_value() && "Expecting valid maxDepth");
+  if (maxDepth != 1)
+    return false;
+
+  unsigned numCandidateArgs = count_if(func.getArguments(), isCandidateArg);
+  if (numCandidateArgs < 2) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "not a candidate: not enough candidate accessors\n");
+    return false;
+  }
+  if (numCandidateArgs > functionSpecializationAccessorLimit) {
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << "not a candidate: exceed accessor limit\n");
+    return false;
+  }
+
+  LLVM_DEBUG(llvm::dbgs().indent(2)
+             << "Found candidate: " << func.getName() << "\n");
+  return true;
+}
+
+bool FunctionSpecializationPass::isCandidateAccessorPair(
+    AccessorPtrType acc1, AccessorPtrType acc2) const {
+  if (acc1 == acc2)
+    return false;
+
+  if (!relaxedAliasing) {
+    auto acc1Ty = cast<sycl::AccessorType>(acc1.getType().getElementType());
+    auto acc2Ty = cast<sycl::AccessorType>(acc2.getType().getElementType());
+    if (acc1Ty.getType() != acc2Ty.getType())
+      return false;
+  }
+
+  return true;
```

Comment:
I may not understand the code here. If the input types are the same, shouldn't we always check for aliasing? Anyway, I think refactoring to reuse code [here](https://github.com/intel/llvm/blob/sycl-mlir/mlir-sycl/lib/Dialect/SYCL/Analysis/AliasAnalysis.cpp#L93) is the way to go.

Reply:
> If the input types are the same, shouldn't we always check for aliasing?

Yes, and we return true for that case.